src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupLoggingHelper.cs

```diff
@@ -93,55 +91,92 @@ internal static string GetParameterText(string prefix, string parameterName, ILi
         /// First line is already indented.
         /// Indent of any subsequent line should be 12 spaces.
         /// </summary>
-        internal static string GetStringFromParameterValue(object parameterValue)
+        internal static string GetStringFromParameterValue(object parameterValue, bool includeMetadata = false)
         {
-            var type = parameterValue.GetType();
-
-            ErrorUtilities.VerifyThrow(!type.IsArray, "scalars only");
-
-            if (type == typeof(string))
+            using (var sb = new ReuseableStringBuilder())
             {
-                return (string)parameterValue;
+                AppendStringFromParameterValue(sb, parameterValue, includeMetadata);
+                return sb.ToString();
             }
-            else if (type.GetTypeInfo().IsValueType)
+        }
+
+        // Avoid allocating a temporary list to hold metadata for sorting every time.
+        // Each thread gets its own copy.
+        [ThreadStatic]
+        private static List<KeyValuePair<string, string>> keyValuePairList;
+
+        private static void AppendStringFromParameterValue(ReuseableStringBuilder sb, object parameterValue, bool includeMetadata = false)
+        {
+            if (parameterValue is string text)
             {
-                return (string)Convert.ChangeType(parameterValue, typeof(string), CultureInfo.CurrentCulture);
+                sb.Append(text);
             }
-            else if (typeof(ITaskItem).GetTypeInfo().IsAssignableFrom(type.GetTypeInfo()))
+            else if (parameterValue is ITaskItem item)
             {
-                var item = ((ITaskItem)parameterValue);
-                string result = item.ItemSpec;
-
-                var customMetadata = item.CloneCustomMetadata();
+                sb.Append(item.ItemSpec);
 
-                if (customMetadata.Count > 0)
+                if (includeMetadata)
                 {
-                    result += "\n";
-                    var names = new List<string>();
+                    var customMetadata = item.CloneCustomMetadata();
```

Comment:
Do we need to clone the metadata or can we just access it?

Reply:
The original code had CloneCustomMetadata() and I think it's the blessed way to retrieve a thread-safe read-only snapshot of custom metadata only.

Reply:
After we've reached this point, the custom metadata won't change, right? In which case, since we're also not modifying it, does it need to be thread-safe or read-only?